libcudacxx/include/cuda/std/__iterator/counted_iterator.h

```diff
@@ -0,0 +1,456 @@
+  }
+
+  _CCCL_TEMPLATE(class _I2 = _Iter)
+  _CCCL_REQUIRES(random_access_iterator<_I2>)
+  _LIBCUDACXX_HIDE_FROM_ABI constexpr counted_iterator& operator-=(iter_difference_t<_I2> __n)
+  {
+    _CCCL_ASSERT(-__n <= __count_,
+                 "Attempt to subtract too large of a size: "
+                 "counted_iterator would be decremented before the "
+                 "first element of its range.");
+    __current_ -= __n;
+    __count_ += __n;
+    return *this;
+  }
+
+  _CCCL_TEMPLATE(class _I2 = _Iter)
+  _CCCL_REQUIRES(random_access_iterator<_I2>)
+  [[nodiscard]] _LIBCUDACXX_HIDE_FROM_ABI constexpr decltype(auto) operator[](iter_difference_t<_I2> __n) const
+  {
+    _CCCL_ASSERT(__n < __count_, "Subscript argument must be less than size.");
+    return __current_[__n];
+  }
+
+  _CCCL_TEMPLATE(class _I2)
+  _CCCL_REQUIRES(common_with<_I2, _Iter>)
+  _CCCL_NODISCARD_FRIEND _LIBCUDACXX_HIDE_FROM_ABI constexpr bool
+  operator==(const counted_iterator& __lhs, const counted_iterator<_I2>& __rhs) noexcept
+  {
+    return __lhs.__count_ == __rhs.__count_;
+  }
+
+#if _CCCL_STD_VER <= 2017
+  _CCCL_NODISCARD_FRIEND _LIBCUDACXX_HIDE_FROM_ABI constexpr bool
+  operator==(const counted_iterator& __lhs, const counted_iterator& __rhs) noexcept
+  {
+    return __lhs.__count_ == __rhs.__count_;
+  }
+
+  _CCCL_NODISCARD_FRIEND _LIBCUDACXX_HIDE_FROM_ABI constexpr bool
+  operator!=(const counted_iterator& __lhs, const counted_iterator& __rhs) noexcept
+  {
+    return __lhs.__count_ != __rhs.__count_;
+  }
+
+  _CCCL_TEMPLATE(class _I2)
+  _CCCL_REQUIRES(common_with<_I2, _Iter>)
+  _CCCL_NODISCARD_FRIEND _LIBCUDACXX_HIDE_FROM_ABI constexpr bool
+  operator!=(const counted_iterator& __lhs, const counted_iterator<_I2>& __rhs) noexcept
+  {
+    return __lhs.__count_ != __rhs.__count_;
+  }
+#endif // _CCCL_STD_VER <= 2017
+
+  _CCCL_NODISCARD_FRIEND _LIBCUDACXX_HIDE_FROM_ABI constexpr bool
+  operator==(const counted_iterator& __lhs, default_sentinel_t) noexcept
+  {
+    return __lhs.__count_ == 0;
+  }
+
+#if _CCCL_STD_VER <= 2017
+  _CCCL_NODISCARD_FRIEND _LIBCUDACXX_HIDE_FROM_ABI constexpr bool
+  operator==(default_sentinel_t, const counted_iterator& __lhs) noexcept
+  {
+    return __lhs.__count_ == 0;
+  }
+
+  _CCCL_NODISCARD_FRIEND _LIBCUDACXX_HIDE_FROM_ABI constexpr bool
+  operator!=(const counted_iterator& __lhs, default_sentinel_t) noexcept
+  {
+    return __lhs.__count_ != 0;
+  }
+
+  _CCCL_NODISCARD_FRIEND _LIBCUDACXX_HIDE_FROM_ABI constexpr bool
+  operator!=(default_sentinel_t, const counted_iterator& __lhs) noexcept
+  {
+    return __lhs.__count_ != 0;
+  }
+#endif // _CCCL_STD_VER <= 2017
+
+#if _LIBCUDACXX_HAS_SPACESHIP_OPERATOR()
+  _CCCL_TEMPLATE(class _I2)
+  _CCCL_REQUIRES(common_with<_I2, _Iter>)
+  _CCCL_NODISCARD_FRIEND _LIBCUDACXX_HIDE_FROM_ABI constexpr strong_ordering
+  operator<=>(const counted_iterator& __lhs, const counted_iterator<_I2>& __rhs) noexcept
+  {
+    return __lhs.__count_ <=> __rhs.__count_;
```

Comment:
i think this should be:
```suggestion
    return __rhs.__count_ <=> __lhs.__count_;
```

see https://eel.is/c++draft/counted.iter.cmp#6. likewise for the other relational ops.